### PR TITLE
Sample Triplanar & Calculate LevelOfDetail Texture2D Node

### PR DESCRIPTION
**Pull request by bencloward**
### Purpose of this PR
Created the new version of the Triplanar Node

Added a new 2 Sample mode that does the triplanar projection using only two texture samples instead of 3.  Once we have working shaders, I need to test this to see if it saves performance vs the 3 sample method.

Created the CalculateLevelOfDetailTexture2D node

---
### Comments to reviewers
I changed the name of the node to Sample Triplanar and moved it to the Input/Texture category both in the menu and in the physical location of the file.  I did this because the node doesn't really fit in the UV category since it's not producing UV coordinates.  We have another node - the UVTriplanar node - that creates UVs for doing your own triplanar projects, so this change also helps to differentiate the two from each other.

### CODE REVIEW

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+        public static NodeDescriptor NodeDescriptor => new(
+            Version,
+            Name,
+            new FunctionDescriptor[] {
+                new(
+                    1,
+                    "SampleTriplanarDefault",
+@"  Node_UV = Position * Tile;
+    Node_Blend = SafePositivePow_float(Normal, min(Blend, floor(log2(Min_float())/log2(1/sqrt(3)))) );
```
Comment (tracychen1234) [inline com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs:19]: Looks like the shader will switch between SafePositivePow_float and [SafePositivePow_half](https://github.com/Unity-Technologies/Graphics/blob/216a86eca4214d7f1aa009649a7967e711b05ac9/Packages/com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl#L685) base on the precision, not sure what will be a good way to do here tho🤔

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    Node_Blend /= dot(Node_Blend, 1.0);
+    Node_X = SAMPLE_TEXTURE2D(UnityBuildTexture2DStructNoScale(Texture).tex, Sampler.samplerstate, Node_UV.zy);
```
Comment (tracychen1234) [inline com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs:21]: Looks like the `UnityBuildTexture2DStructNoScale()` will be automatically added when compiling the shader, I think `Texture.tex` should be good here

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                ),
+                new(
+                    1,
+                    "SampleTriplanarNormal",
+@"  Node_UV = Position * Tile;
+    Node_Blend = SafePositivePow_float(Normal, min(Blend, floor(log2(Min_float())/log2(1/sqrt(3)))) );
+    Node_Blend /= dot(Node_Blend, 1.0);
+    Node_X = UnpackNormal(SAMPLE_TEXTURE2D(UnityBuildTexture2DStructNoScale(Texture).tex, Sampler.samplerstate, Node_UV.zy));
+    Node_Y = UnpackNormal(SAMPLE_TEXTURE2D(UnityBuildTexture2DStructNoScale(Texture).tex, Sampler.samplerstate, Node_UV.xz));
+    Node_Z = UnpackNormal(SAMPLE_TEXTURE2D(UnityBuildTexture2DStructNoScale(Texture).tex, Sampler.samplerstate, Node_UV.xy));
```
Comment (tracychen1234) [inline com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs:50]: Same here

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	//create coordinates
+	xCoords.x = Position[ma.y]; xCoords.y = Position[ma.z];
+	xDDX.x = dpdx[ma.y]; xDDX.y = dpdx[ma.z];
+	xDDY.x = dpdy[ma.y]; xDDY.y = dpdy[ma.z];
+	yCoords.x = Position[me.y]; yCoords.y = Position[me.z];
+	yDDX.x = dpdx[me.y]; yDDX.y = dpdx[me.z];
+	yDDY.x = dpdy[me.y]; yDDY.y = dpdy[me.z];	
+	// project+fetch
+	Node_X = SAMPLE_TEXTURE2D_GRAD(UnityBuildTexture2DStructNoScale(Texture).tex, Sampler.samplerstate, xCoords, xDDX, xDDY);
+	Node_Y = SAMPLE_TEXTURE2D_GRAD(UnityBuildTexture2DStructNoScale(Texture).tex, Sampler.samplerstate, yCoords, yDDX, yDDY);
```
Comment (tracychen1234) [inline com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs:105]: same here

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                    new ParameterDescriptor("yDDY", TYPE.Vec2, Usage.Local)
+                )
+            }
+        );
+
+        public static NodeUIDescriptor NodeUIDescriptor => new(
+            Version,
+            Name,
+            tooltip: "Samples a texture three times and projects in front/back, top/bottom, and left/right.",
+            categories: new string[2] { "Input", "Texture" },
```
Comment (tracychen1234) [inline com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs:141]: In sg1 it's under `UV` , do we want to move it to `Input`>`Texture`?

**Review by tracychen1234 — CHANGES_REQUESTED** [review]

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs (hunk 1 above)
Comment (bencloward) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs:19]: I don't know how to fix it either.

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs (hunk 5 above)
Comment (bencloward) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs:141]: I think we should move it, yes.  Do you? It doesn't create UV coordinates as output like the other UV nodes do.  It samples textures like the other Input/Texture nodes.  We're adding the UVTriplanar node to the UV category, so moving this one helps to differentiate them.

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs (hunk 2 above)
Comment (bencloward) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs]: Fixed - thank you!

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs (hunk 3 above)
Comment (bencloward) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs]: Fixed

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs (hunk 4 above)
Comment (bencloward) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs]: Fixed

**Comment by bencloward** [comment]
Could you also review the Calculate Level of Detail Texture 2D node that I added to the review?  Sorry to add while you were already reviewing the other one.  :P

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/CalculateLevelOfDetailTexture2DNode.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        public static string Name = "CalculateLevelOfDetailTexture2D";
+        public static int Version = 1;
+        public static FunctionDescriptor FunctionDescriptor => new(
+            Version,
+            Name,
+@"#if (SHADER_TARGET >= 41)
+    if (Clamp)
+    {
+        LOD = UnityBuildTexture2DStructNoScale(Texture).tex.CalculateLevelOfDetail(Sampler.samplerstate, UV);
```
Comment (tracychen1234) [inline com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/CalculateLevelOfDetailTexture2DNode.cs:16]: same here

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs (hunk 5 above)
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs:141]: sounds good!

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs (hunk 1 above)
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs:19]: we can add a comment here and come back later when we have a way to fix it?

Comment (bencloward) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/SampleTriplanarNode.cs:20]: Done

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/CalculateLevelOfDetailTexture2DNode.cs (hunk 6 above)
Comment (bencloward) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Input/Textures/CalculateLevelOfDetailTexture2DNode.cs]: Fixed

**Review by tracychen1234 — APPROVED** [review]

**Review by brett-groshong — APPROVED** [review]

**Pull request merged**